### PR TITLE
Use PSyIR to identify NemoKernel (#361)

### PR DESCRIPTION
**Pull request by arporter**
This PR improves the way we identify NemoKernel objects by analysing the PSyIR rather than the fparser2 parse tree. I've taken the opportunity to tidy-up the `NemoKernel` class to remove unused attributes.

Currently several tests have been marked as xfail because they need PSyIR support for If Constructs (#347).

### CODE REVIEW

**Comment by arporter** [comment]
Note to self: need to correct parent information in PSyIR nodes that are moved into the `KernelSchedule`

**Comment by arporter** [comment]
Now that support for If/Case statements is on master, I've brought this branch up-to-date and removed the x-failing annotation previous added to a couple of tests. I've also taken the opportunity to update the NEMO-related documentation a little. Ready for first review from @rupertford.

File: doc/examples.rst [hunk 1]
```diff
@@ -190,3 +190,27 @@ to the specified Kernel:
...
+Example 1: OpenMP parallelisation of tra_adv
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Demonstrates the use of PSyclone to parallelise the loops over vertical levels
+in some NEMO tracer-advection code using OpenMP.
+
+Example 2: OpenMP parallelisation of traldf_iso
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/examples.rst]: Oh yes, don't know what happened there.

File: doc/examples.rst [hunk 2]
```diff
@@ -190,3 +190,27 @@ to the specified Kernel:
...
+Demonstrates the use of PSyclone to parallelise the loops over vertical levels
+in some NEMO tracer-diffusion code using OpenMP.
+
+Example 3: OpenACC parallelisation of tra_adv
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Demonstrates the introduction of simple OpenACC parallelisation (using the
+``data`` and ``kernels`` directives) for a tracer advection benchmark.
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/examples.rst]: Good point, done.

File: doc/nemo.rst [hunk 3]
```diff
@@ -209,22 +209,18 @@ we list the current, known limitations/issues:
  3. All recognised loops (levels, latitude etc.) are assumed to be
     parallelisable. This will not always be the case (e.g. tridiagonal
     solve has a loop-carried dependence in the vertical);
- 4. Loops/kernels within CASE statements are not found;
- 5. Labelled do-loops are not handled (i.e. they will probably end up
+ 4. Labelled do-loops are not handled (i.e. they will probably end up
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/nemo.rst]: Just me being sloppy I think. I've changed it to 'will.'

File: src/psyclone/nemo.py [hunk 4]
```diff
@@ -90,13 +90,12 @@ def _create_child(self, child, parent=None):
         '''
         if NemoLoop.match(child):
             return NemoLoop(child, parent=parent)
-        elif isinstance(child, Fortran2003.Nonlabel_Do_Stmt):
-            pass
-        elif NemoImplicitLoop.match(child):
+        if isinstance(child, Fortran2003.Nonlabel_Do_Stmt):
+            return None
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/nemo.py]: Fixed.

[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/nemo.py]: Done.

File: src/psyclone/nemo.py [hunk 5]
```diff
@@ -295,183 +294,62 @@ def __str__(self):
...
         self._parent = parent
+        # The corresponding set of nodes in the fparser2 parse tree
+        self._ast = parse_tree
+        # Create a kernel schedule
+        self._kern_schedule = KernelSchedule(self._name)
+        # Attach the PSyIR sub-tree to it
+        self._kern_schedule.children = psyir_nodes[:]
+        # Update the parent info for each node we've moved
+        for node in self._kern_schedule.children:
+            node.parent = parent
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/nemo.py]: Oops. Well spotted.

File: src/psyclone/nemo.py [hunk 6]
```diff
@@ -295,183 +294,62 @@ def __str__(self):
...
-
-        # Currently a kernel cannot contain implicit loops.
-        # TODO we may have to differentiate between implicit loops over
-        # grid points and any other implicit loop. Possibly using the
-        # scope of the array being accessed?
-        impl_loops = walk_ast(node.content, [Subscript_Triplet])
-        if impl_loops:
+        from psyclone.psyGen import CodeBlock
+        if node.walk(node.children, (CodeBlock, NemoLoop, NemoImplicitLoop)):
+            # A kernel cannot contain other loops, reads or writes or calls
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/nemo.py]: Yes, the code has evolved but the comment hadn't. I've tidied it up and also removed `NemoImplicitLoop` from the list as it is a sub-class of `NemoLoop`.

File: src/psyclone/nemo.py [hunk 7]
```diff
@@ -295,183 +294,62 @@ def __str__(self):
...
 
     @staticmethod
     def match(node):
         '''
-        Whether or not the AST fragment pointed to by node represents a
+        Whether or not the PSyIR sub-tree pointed to by node represents a
         kernel. A kernel is defined as a section of code that sits
         within a recognised loop structure and does not itself contain
-        loops or IO operations.
+        loops, subroutine calls or IO operations.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/nemo.py]: Fixed.

File: src/psyclone/nemo.py [hunk 8]
```diff
@@ -556,12 +433,13 @@ def __init__(self, ast, parent=None):
...
-            self.addchild(NemoKern(self._ast, parent=self))
-            return
-        # It's not - walk on down the AST...
+        # First process the rest of the parse tree below this point
         self.process_nodes(self, self._ast.content, self._ast)
+        # Now check the PSyIR of this loop body to see whether it is
+        # a valid kernel
+        if NemoKern.match(self):
+            # It is, so we create a new kernel object
+            kernel = NemoKern(self.children, self._ast, parent=self)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/nemo.py]: Yes, but this highlights something that I've just spotted and am not very happy with - the `NemoKern` constructor takes it upon itself to modify the state of `parent`. At the time it seemed like a useful way of doing it but now it feels wrong (breaks encapsulation) so I shall change this.

File: src/psyclone/nemo.py [hunk 9]
```diff
@@ -295,183 +294,62 @@ def __str__(self):
...
+    no self._children but they do have a KernelSchedule.
+
+    :param psyir_nodes: the list of PSyIR nodes that represent the body \
+                        of this kernel.
+    :type psyir_nodes: list of :py:class:`psyclone.psyGen.Node`
+    :param parse_tree: Reference to the innermost loop in the fparser2 parse \
+                       tree that encloses this kernel.
+    :type parse_tree: \
+              :py:class:`fparser.two.Fortran2003.Block_Nonlabel_Do_Construct`
+    :param parent: the parent of this Kernel node in the PSyIR.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/nemo.py]: Fixed.

File: src/psyclone/tests/nemo/nemo_psy_test.py [hunk 10]
```diff
@@ -228,6 +228,27 @@ def test_io_not_kernel():
     assert len(kerns) == 1
 
 
+def test_fn_call_no_kernel(parser):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/nemo/nemo_psy_test.py]: Good point. Done.

[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/nemo/nemo_psy_test.py]: Also a good point. I've added one and tidied for some other pylint errors.

File: src/psyclone/tests/nemo/transformations/openacc/data_directive_test.py [hunk 11]
```diff
@@ -282,12 +284,15 @@ def test_data_ref():
                            api=API, line_length=False)
     psy = PSyFactory(API, distributed_memory=False).create(invoke_info)
     schedule = psy.invokes.get('data_ref').schedule
+    schedule.view()
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/nemo/transformations/openacc/data_directive_test.py]: Done.

**Comment by arporter** [comment]
All reviewer requests have been addressed. Just waiting on the coverage report from CodeCov.

**Comment by arporter** [comment]
Coverage of the diff is now 100%. Ready for 2nd review.

File: src/psyclone/psyGen.py [hunk 12]
```diff
@@ -5576,6 +5571,7 @@ def _if_stmt_handler(self, node, parent):
         :rtype: :py:class:`psyclone.psyGen.IfBlock`
         '''
         ifblock = IfBlock(parent=parent, annotation='was_single_stmt')
+        ifblock.ast = node
```
Comment (rupertford) [inline src/psyclone/psyGen.py]: I presume this is a bug fix.

**Comment by rupertford** [comment]
Back to @arporter

File: src/psyclone/tests/nemo/transformations/loops/loop_test.py [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+            "      DO ji = 1, jpi, 1\n"
+            "        umask(ji, jj, jk) = 0.0D0\n"
+            "      END DO\n"
+            "    END DO\n"
+            "  END DO\n" in gen_code)
+
+
+@pytest.mark.xfail(reason="Code being transformed already declares ji and jj "
+                   "and so we get duplicate declarations. Needs symbol "
+                   "table - #255.")
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/nemo/transformations/loops/loop_test.py]: Oh yes, well spotted. I've created a new Issue and referenced that instead.

File: doc/nemo.rst [hunk 14]
```diff
@@ -209,22 +209,18 @@ we list the current, known limitations/issues:
...
+ 5. Loops are currently only permitted to contain one kernel.  This
     restriction will have to be lifted in order to permit loop fusion;
- 8. Array slices with specified bounds (e.g. umask(1:10)) are not yet
+ 6. Array slices with specified bounds (e.g. umask(1:10)) are not yet
     supported and will raise a TransformationError when attempting to
     transform them into explicit loops;
- 9. When generating new variable names, no attempt is made to avoid
+ 7. When generating new variable names, no attempt is made to avoid
     clashing with variables already present in the NEMO source. This
     will be resolved once a symbol table is available (#255).
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/nemo.rst]: Changed to point to #381

File: src/psyclone/psyGen.py (hunk 12 above)
Comment (arporter) [reply src/psyclone/psyGen.py]: Yes - we weren't setting the `._ast` member of IfBlock previously and that causes problems when walking the tree (in transformations).

**Pull request merged**